tests/test_tcp_models.py

```diff
@@ -0,0 +1,381 @@
+    """Test request from an Inverter - TCP source."""
+    serial_number = 987654321
+
+    (server_exit, port) = tcp_server(serial_number, "tcp_reply.data")
+
+    client = OmnikInverter(
+        host="localhost",
+        source_type="tcp",
+        serial_number=serial_number,
+        tcp_port=port,
+    )
+
+    inverter: Inverter = await client.inverter()
+
+    assert inverter
+    assert inverter.solar_rated_power is None
+    assert inverter.solar_current_power == 2615
+
+    assert inverter.model is None
+    assert inverter.serial_number == "NLDN012345CS4321"
+    assert inverter.temperature == 43.1
+    assert inverter.dc_input_voltage == [187.3, 188.9, None]
+    assert inverter.dc_input_current == [8.1, 7.800000000000001, None]
+    assert inverter.ac_output_voltage == [239.60000000000002, None, None]
+    assert inverter.ac_output_current == [10.8, None, None]
+    assert inverter.ac_output_frequency == [50.06, None, None]
+    assert inverter.ac_output_power == [2615.0, None, None]
+    assert inverter.solar_energy_today == 7.4
+    assert inverter.solar_energy_total == 65432.100000000006
+    assert inverter.solar_hours_total == 54321
+    assert inverter.inverter_active is True
+    assert inverter.firmware == "NL1-V1.0-0077-4"
+    assert inverter.firmware_slave == "V2.0-0024"
+
+    await server_exit
+
+
+async def test_inverter_tcp_offline() -> None:
+    """Test request from an Inverter (offline) - TCP source."""
+    serial_number = 1608449224
+
+    (server_exit, port) = tcp_server(serial_number, "tcp_reply_offline.data")
+
+    client = OmnikInverter(
+        host="localhost",
+        source_type="tcp",
+        serial_number=serial_number,
+        tcp_port=port,
+    )
+
+    inverter: Inverter = await client.inverter()
+
+    assert inverter
+    assert inverter.solar_rated_power is None
+    assert inverter.solar_current_power == 0
+
+    assert inverter.model is None
+    assert inverter.serial_number == "NLBN4020157P9024"
+    assert inverter.temperature is None
+    assert inverter.dc_input_voltage == [0.0, 0.0, 0.0]
+    assert inverter.dc_input_current == [0.0, 0.0, 0.0]
+    assert inverter.ac_output_voltage == [0.0, 0.0, 0.0]
+    assert inverter.ac_output_current == [0.0, 0.0, 0.0]
+    assert inverter.ac_output_frequency == [0.0, 0.0, 0.0]
+    assert inverter.ac_output_power == [0.0, 0.0, 0.0]
+    assert inverter.solar_energy_today == 4.7
+    assert inverter.solar_energy_total == 15818.0
+    assert inverter.solar_hours_total == 0
+    assert inverter.inverter_active is False
+    assert not inverter.firmware
+    assert not inverter.firmware_slave
+
+    await server_exit
+
+
+async def test_connection_broken() -> None:
+    """Test closed connection after successful handshake - TCP source."""
+    serial_number = 1
+
+    def close_immediately(conn: socket) -> None:
+        """Close the connection immediately without _reading_ nor writing."""
+        conn.close()
```

Comment:
This test is a bit flaky (by nature) and sometimes doesn't see an error raised anywhere (i.e. not in `wait_closed()`, nor being confused by no bytes being consumed from the stream), instead processing an empty `raw_msg` and failing with:

```
omnikinverter.exceptions.OmnikInverterPacketInvalidError: None of the messages contained an information reply!
```

Reply:
As I have no idea how to properly fix this (cleanly `shutdown(SHUT_RDWR)`'ing this socket allows me to never even see the `except OSError as exception` we're testing for here) we could possibly ignore this low-level exception from the tests: it's more so to wrap the low-level "maybe something bad happened on the socket" and give it a nicer message for the user, not important logic-wise.

Reply:
Alternatively I could also insert the following below the `finally:` block to trick the same error, but that results in an obvious coverage issue:

```python
        if not raw_msg:
            msg = "Failed to communicate with the Omnik Inverter device over TCP"
            raise OmnikInverterConnectionError(msg)
```

Reply:
CI also ran into the flake now, and it happens way more often locally too...

Reply:
I think I got a solution for this: according to the internet we can force an `RST` packet (the one that causes the `OSError("Connection reset by peer")` that we re-wrap in `OmnikInverterConnectionError("Failed to communicate with the Omnik Inverter device over TCP")` here) by enabling linger, but with a timeout of 0.

Note that I did have to make sure that we first read from the socket (and still assert that the client sent valid data) otherwise the connection might have been closed so quickly that `open_connection()` raises the `OSError` instead which we bubble up with `"Failed to open a TCP connection to the Omnik Inverter device"` already.